File: 10_25_2016_binary_search_tree_iterator/chang.java
public class BSTIterator {
    //@param root: The root of binary tree.
    private Stack<TreeNode> stack = new Stack<TreeNode>();
    private TreeNode cur;
    
    public BSTIterator(TreeNode root) {
        // 
        cur = root;
    }

    //@return: True if there has next node, or false
    public boolean hasNext() {
        // 
        return (cur != null || !stack.isEmpty());
    }
    
    //@return: return next node
    public TreeNode next() {
        while (cur != null) {
            stack.push(cur);
            cur = cur.left;
        }
        
        TreeNode next = stack.pop();
        cur = next;
        cur = cur.right;
        
        return next;
        
        
    }
}

File: 11_03_2016_Binary_Tree_Path/chang.py
class Solution:
    # @param {TreeNode} root the root of the binary tree
    # @return {List[str]} all root-to-leaf paths
    def binaryTreePaths(self, root):
        result = []
        path = []
        if root is None:
            return []
        self.dfs(path, result, root)
        return result
        
        
    def dfs(self, path, result, node):
        path.append(str(node.val))
        
        if node.left is None and node.right is None:
            result.append("->".join(path))
            path.pop()
            return
        
        if node.left:
            self.dfs(path, result, node.left)
        if node.right:
            self.dfs(path, result, node.right)
        
        path.pop()
            
   
